rmw_fastrtps_shared_cpp/src/rmw_node_info_and_types.cpp

```diff
@@ -0,0 +1,490 @@
+}
+
+rmw_ret_t
+__rmw_get_subscriber_names_and_types_by_node(
+  const char * identifier,
+  const rmw_node_t * node,
+  rcutils_allocator_t * allocator,
+  const char * node_name,
+  const char * node_namespace,
+  bool no_demangle,
+  rmw_names_and_types_t * topic_names_and_types)
+{
+  RetrieveCache retrieve_sub_cache =
+    [](CustomParticipantInfo & participant_info) -> const LockedObject<TopicCache> & {
+      return participant_info.secondarySubListener->topic_cache_;
+    };
+  return __rmw_get_topic_names_and_types_by_node(identifier, node, allocator, node_name,
+           node_namespace, no_demangle, retrieve_sub_cache, topic_names_and_types);
+}
+
+rmw_ret_t
+__rmw_get_publisher_names_and_types_by_node(
+  const char * identifier,
+  const rmw_node_t * node,
+  rcutils_allocator_t * allocator,
+  const char * node_name,
+  const char * node_namespace,
+  bool no_demangle,
+  rmw_names_and_types_t * topic_names_and_types)
+{
+  RetrieveCache retrieve_pub_cache =
+    [](CustomParticipantInfo & participant_info) -> const LockedObject<TopicCache> & {
+      return participant_info.secondaryPubListener->topic_cache_;
+    };
+  return __rmw_get_topic_names_and_types_by_node(identifier, node, allocator, node_name,
+           node_namespace, no_demangle, retrieve_pub_cache, topic_names_and_types);
+}
+
+rmw_ret_t
+__rmw_get_service_names_and_types_by_node(
+  const char * identifier,
+  const rmw_node_t * node,
+  rcutils_allocator_t * allocator,
+  const char * node_name,
+  const char * node_namespace,
+  rmw_names_and_types_t * service_names_and_types)
+{
+  rmw_ret_t valid_input = __validate_input(identifier, node, allocator, node_name,
+      node_namespace, service_names_and_types);
+  if (valid_input != RMW_RET_OK) {
+    return valid_input;
+  }
+  auto impl = static_cast<CustomParticipantInfo *>(node->data);
+  __log_debug_information(*impl);
+
+  GUID_t guid;
+  rmw_ret_t valid_guid = __get_guid_by_name(node, node_name, node_namespace, guid);
+  if (valid_guid != RMW_RET_OK) {
+    return valid_guid;
+  }
+
+  std::map<std::string, std::set<std::string>> services;
+  {
+    auto & topic_cache = impl->secondaryPubListener->topic_cache_;
+    std::lock_guard<std::mutex> guard(topic_cache.getMutex());
+    const auto & node_topics = topic_cache.getParticipantToTopics().find(guid);
+    if (node_topics != topic_cache.getParticipantToTopics().end()) {
+      for (auto & topic_pair : node_topics->second) {
+        std::string service_name = _demangle_service_from_topic(topic_pair.first);
+        if (!service_name.length()) {
+          // not a service
+          continue;
+        }
+        for (auto & itt : topic_pair.second) {
+          std::string service_type = _demangle_service_type_only(itt);
+          if (service_type.length()) {
+            services[service_name].insert(service_type);
+          }
+        }
+      }
+    }
+  }
+
+  if (services.size()) {
```

Comment:
This is a subjective thing, but we could do an early return and lose one level of indentation.  Something like:

```
if (services.empty()) {
  return RMW_RET_OK;
}
```